[tado] Code optimizations

andrewfg opened the pull request:
In this PR we do the following:

1. Remove old authentication code as suggested [here](https://github.com/openhab/openhab-addons/pull/18481#discussion_r2092959633)
2. Remove some other unused code
3. Refactor calls to imported methods that have in the meantime been marked as `@Deprecated`.

https://github.com/openhab/openhab-addons/pull/18481#discussion_r2092959633

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
@jlaur for info. I don't suggest to actually cherry pick this into 4.3.x. But since my operative system is still on 4.3.x I do need to make an 'ad hoc' back port for live testing anyway, and since I am not 100% fluent on that process I shall probably defer that until the weekend. But please feel free to review the code anyway.

lsiepel commented:
> In this PR we do the following:
> 
> 1. Remove old authentication code as suggested [here](https://github.com/openhab/openhab-addons/pull/18481#discussion_r2092959633)
> 2. Remove some other unused code
> 3. Refactor calls to imported methods that have in the meantime been `@Deprecated`.
> 
> [#18481 (comment)](https://github.com/openhab/openhab-addons/pull/18481#discussion_r2092959633)
> 
> Signed-off-by: Andrew Fiddian-Green [software@whitebear.ch](mailto:software@whitebear.ch)

Backporting is simpler and with less risc if the change is limited to removing the old flow. Replacing the deprecated calls should be in a seperate PR

andrewfg commented:
> Backporting is simpler and with less risk if the change is limited 

@lsiepel really I have no intention of back porting this one. The difficulty is not about fixing the calls to `@Deprecated` methods. The issue is that in the interval between the last v4.3.x build we have had three other PRs merged into v5.x that should definitely NOT be back ported to v4.3.x.

However for testing purposes I was able to munge together an amalgam the code from this PR plus the not-to-be-back-ported code from the other three PRs into my own v4.3.x snapshot ("eat your own dog food"), and I can confirm that it is tested and fully running.

![image](https://github.com/user-attachments/assets/b7b82ba2-5c98-409e-b737-bdb21b91ef2a)

![image](https://github.com/user-attachments/assets/93c219ab-5063-4834-904b-068f15fb9b09)

andrewfg commented on bundles/org.openhab.binding.tado/README.md:
```diff
@@ -12,33 +12,30 @@ You can then monitor and control all zone types (Heating, AC, Hot Water) as well
 | Parameter     | Optional | Description                                                                        |
 |---------------|----------|------------------------------------------------------------------------------------|
-| `useRfc8628`  | yes      | Determines if the binding shall use oAuth RFC-8628 authentication                  |
 | `rfcWithUser` | yes      | Determines if the user name shall be included in the oAuth RFC-8628 authentication |
```
Yeah. I thought about that. Let me think about it some more..

andrewfg replied:
Ok. Done.

andrewfg commented:
@lsiepel please don’t merge this yet; the last changes will break existing functionality.

andrewfg commented:
@lsiepel removing the `rfcWithUser` config param (and just relying on a non blank `username`) would have broken the functionality for existing users. Specifically such users having a single tado account will have a `username` config param from the prior authentication scheme. So if we would have changed the functionality it would have forced single tado account users into the multi account mode of authentication. I want to avoid this because a) it is more complex to understand, and b) this PR would  for every single account user make the existing token inaccessible and therefore force them to go through a new authentication process again.

lsiepel commented:
> aram (and just relying on a non blank `username`) would have broken the functionality for existing users. Specifically such users having a single tado account will have a `username` config param from the prior authentication scheme. So if we would have changed the functionality it would have forced single tado account users into the multi account mode of authentication. I w

How would that break ?

1. User has one account:
    1. rfcWithUser is true => the username is required. (but if it was empty, it wouldn't matter)
    2. rfcWithUser is false => the username does not matter.
2. User has multiple acocunts:
    1. a valid configuration would require rfcWithUser true and a username for _all_ accounts.

So whenever the `username` is set it should be used for the tokenstuff, otherwise ignored. What am i missing here?

andrewfg commented:
@lsiepel the migration path is as follows..

1. Before March 2025 the authentication was username/password; all existing things therefore had a `username` defined and used.

2. In #18433 we added support for OAuth; only one account was supported; the token is created/persisted in the 'tado' topic; `username` was retained but not used.

3. In #18481 we added support for multiple accounts; single account things based on existing tokens would continue to work with the token persisted in the 'tado' topic; in that case `username` was still retained but not used; if the user adds another thing / account they must create it with with `rfcWithUser` and respective `username`; for this new thing the token is created/persisted in the 'tado-username' topic.

4. In #18691 any existing username config params are still retained; and any existing tokens remain persisted in the same place(s) as before.

4a. my version: single- (or first-) account things based on existing tokens will continue to work with the token created/persisted in the 'tado' topic, and any second- account things will continue to work with the tokens persisted in the respective 'tado-username' topic; if the user adds another account thing its token is created/persisted in the respective 'tado-username' topic.

4b. your version: single- (or first-) account things based on existing tokens will have the retained (but temporarily not used) `username`; and since `username` is not empty, these things will NOT continue to work with the token created/persisted in the 'tado' topic; instead the binding tries to retrieve a token from the 'tado-username' topic -- which does not exist; so the existing valid token becomes 'orphaned' in the 'tado' topic; and the user is forced to do a reauthentication so that the created token is now persisted in the 'tado-username' topic; (or the user is required to manually edit the credentials persistence json file); these actions are neither good nor trivial..

andrewfg commented:
@lsiepel from my perspective this one is ready to merge.

lsiepel commented:
One last question regarding the tokens. 

Would it be possible to get a list of cached keys/tokens within the `tado` namespace / prefix?
If so you could detect and copy the token and 0% of the users would need to re authenticated and the configuration would be intuitive and minimal. Just not sure if that is possible.

lsiepel commented:
> . If at the instant that this PR is first installed there is definitely only one thing, then we can indeed safely auto migrate the token from 'tado' => 'tado+first_and_only_username'. But a problem could arise in a multi thing scenario because the sequence of thing instantiation would be indeterminate and the "second" thing might be instantiated before the "first" thing; so we might wrongly move the token from 'tado' => '

Oke, let's abort this scenario. You did more then you could do here. From a distance it looked so simple to 'correct' all configurations. Whenever there is a future breaking change, we should do this simplification.

lsiepel reviewed (approved):
Thanks, LGTM
